dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfigurationService.java

```diff
@@ -22,9 +30,17 @@
  */
 public class DiscoveryConfigurationService {
 
+    private static final Logger log = LogManager.getLogger();
+
     private Map<String, DiscoveryConfiguration> map;
     private Map<Integer, List<String>> toIgnoreMetadataFields = new HashMap<>();
 
+    /**
+     * Discovery configurations, cached by DSO UUID. When a DSO doesn't have its own configuration, we take the one of
+     * the first parent that does. This cache ensures we don't have to go up the hierarchy every time.
+     */
+    private final Map<UUID, DiscoveryConfiguration> uuidMap = new HashMap<>();
```

Comment:
the entry in this second cache is just a reference to the DiscoveryConfiguration so there is no memory overhead due to the DiscoveryConfiguration. Nevertheless, such cache is problematic as a full reindex would result in a large map with the uuid of all the dspace object in memory. Moreover, if an object is moved changing its parent the map will become inconsistent (an item moved to a different collection). Without such a cache the indexing performance would decrease drastically so we need to found a more sophisticated mechanism such as a Thread bound cache limiting the entry to only communities and collections uuid (reusable).

Reply:
For "deep" Comm/Coll hierarchies the performance impact of this cache is pretty significant when searching, and likely even more so during indexing... so I don't think we can get rid of it entirely.

However, if we limit it to Community/Collection only there I don't think there is as much to worry about 
* inconsistency → Communities/Collections don't support moves/mapping like Items do
* memory → Comm/Coll hierarchy is not likely to be large enough for this to be a concern

If consistency is still a concern we can also move the cache out to a separate service with a public method to clear it. Then we can play it safe and start from scratch whenever the Comm/Coll hierarchy is changed (on create, delete, ...)

Another approach we could take would be to look further into the performance of recursively retrieving all of a DSO's parents -- this is likely our bottleneck, if it can be improved the cache may become less crucial. This is something we'd have to look at anyway if we'd consider doing away with Solr-indexed authorizations for #8644.

All in all I don't think we need the additional complexity of a thread-bound cache here.

Reply:
I see this cache has been updated to only include Communities and Collections, and this approach seems better.  So, I think this looks good now. Thanks @ybnd